packages/flutter/lib/src/widgets/actions.dart

```diff
@@ -120,48 +294,55 @@ typedef OnInvokeCallback = void Function(FocusNode node, Intent tag);
 ///    and allows redefining of actions for its descendants.
 ///  * [ActionDispatcher], a class that takes an [Action] and invokes it using a
 ///    [FocusNode] for context.
-class CallbackAction extends Action {
-  /// A const constructor for an [Action].
+class CallbackAction<T extends Intent> extends Action<T> {
+  /// A constructor for a [CallbackAction].
   ///
   /// The `intentKey` and [onInvoke] parameters must not be null.
   /// The [onInvoke] parameter is required.
-  const CallbackAction(LocalKey intentKey, {@required this.onInvoke})
-      : assert(onInvoke != null),
-        super(intentKey);
+  CallbackAction({@required this.onInvoke}) : assert(onInvoke != null);
 
   /// The callback to be called when invoked.
   ///
   /// Must not be null.
   @protected
-  final OnInvokeCallback onInvoke;
+  final OnInvokeCallback<T> onInvoke;
 
   @override
-  void invoke(FocusNode node, Intent intent) => onInvoke.call(node, intent);
+  Object invoke(covariant T intent) => onInvoke.call(intent);
 }
 
-/// An action manager that simply invokes the actions given to it.
+/// An action dispatcher that simply invokes the actions given to it.
+///
+/// See also:
+///
+///  - [ShortcutManager], that uses this class to invoke actions.
+///  - [Shortcuts] widget, which defines key mappings to [Intent]s.
+///  - [Actions] widget, which defines a mapping between a [Key] and an [Action].
 class ActionDispatcher with Diagnosticable {
-  /// Const constructor so that subclasses can be const.
+  /// Const constructor so that subclasses can be immutable.
   const ActionDispatcher();
 
   /// Invokes the given action, optionally without regard for the currently
   /// focused node in the focus tree.
   ///
-  /// Actions invoked will receive the given `focusNode`, or the
-  /// [FocusManager.primaryFocus] if the given `focusNode` is null.
-  ///
-  /// The `action` and `intent` arguments must not be null.
+  /// The action will be invoked with the given `context`, if given, but only if
+  /// the action is a [ContextAction] subclass. If the action isn't a
+  /// [ContextAction] subclass, then the `context` may still be used to ask the
+  /// intent if it is enabled.
   ///
   /// Returns true if the action was successfully invoked.
-  bool invokeAction(Action action, Intent intent, {FocusNode focusNode}) {
+  Object invokeAction(covariant Action<Intent> action, covariant Intent intent, [BuildContext context]) {
     assert(action != null);
     assert(intent != null);
-    focusNode ??= primaryFocus;
-    if (action != null && intent.isEnabled(focusNode.context)) {
-      action.invoke(focusNode, intent);
-      return true;
+    context ??= primaryFocus?.context;
+    if (context == null || action.enabled) {
```

Comment:
So, if no context is provided we always invoke the action even if it claims that it is not enabled? Is that intended or a bug? Seems surprising to me and if intended should at least be documented on the Action's invoke method.